### ISSUE TITLE
Persistent State for `HashMap::compute`

### ISSUE THREAD

**Issue by ibraheemdev** [ibraheemdev/papaya#9]
Currently `HashMap::compute` requires an `FnMut` and has no way of persisting state. For example, inserting a `V` requires cloning if the operation fails and is retried. Internally we memoize consecutive inserts to avoid some of these cases, but that's somewhat of a hack that shouldn't be exposed to users. Instead we might want to provide an `on_failure(Operation<V>)` hook to allow state to be reset, or maybe even add an `Operation::Reset` variant. However, this becomes more complicated for updates and I don't want the API to become convoluted, so maybe `update_with(key, initial, |val| initial.val = f(val))` or `compute_with` would be appropriate along side the lower-level `compute` API.

**Comment by spaolacci** [comment]
fwiw, that's a situation I also faced while porting some code to Papaya. Not taking an `FnOnce` (understandably) also prevent other possibly larger/expensive elements, like a large new value for example, to be moved within the closure. In turn, this means, sometimes, more copy/clone/work has to be done within the closure than strictly required.

An alternative could be, in case of retries, to recycle (reinject) previous outputs into subsequent calls (previously built value for `update` for example).

**Comment by ibraheemdev** [comment]
> An alternative could be, in case of retries, to recycle (reinject) previous outputs into subsequent calls (previously built value for update for example).

Yeah I think something like this would be ideal. I'll try to take some time and design a nice API.
